subsys/bootloader/bl_monotonic_counters/bl_monotonic_counters.c

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <bl_monotonic_counters.h>
+#include <string.h>
+#include <errno.h>
+#include <nrf.h>
+#include <assert.h>
+#include <nrfx_nvmc.h>
+
+#ifdef CONFIG_SECURE_BOOT_STORAGE
+#include "bl_storage.h"
+#else
+#include <pm_config.h>
+#endif
+
+/** Get the counter_collection data structure in the provision data. */
+static const struct counter_collection *get_counter_collection(void)
+{
+#if defined(CONFIG_SECURE_BOOT_STORAGE)
+	struct counter_collection *collection = (struct counter_collection *)
+		&BL_STORAGE->key_data[num_public_keys_read()];
+#else
+	static const struct counter_collection *collection = (struct counter_collection *)PM_PROVISION_ADDRESS;
+#endif
+
+	return nrfx_nvmc_otp_halfword_read((uint32_t)&collection->type) == TYPE_COUNTERS
+		? collection : NULL;
+}
+
+
+/** Get one of the (possibly multiple) counters in the provision data.
+ *
+ *  param[in]  description  Which counter to get. See COUNTER_DESC_*.
+ */
+static const struct monotonic_counter *get_counter_struct(uint16_t description)
+{
+	const struct counter_collection *cnt_collection = get_counter_collection();
+
+	if (cnt_collection == NULL) {
+		return NULL;
+	}
+
+	const struct monotonic_counter *current = cnt_collection->counters;
+
+	for (size_t i = 0; i < nrfx_nvmc_otp_halfword_read((uint32_t)&cnt_collection->num_counters); i++) {
+
+		if (nrfx_nvmc_otp_halfword_read((uint32_t)&current->description) == description) {
+			return current;
+		}
+
+		uint16_t num_slots = nrfx_nvmc_otp_halfword_read((uint32_t)&current->num_counter_slots);
+		current = (const struct monotonic_counter *)
+					&current->counter_slots[num_slots];
+	}
+	return NULL;
+}
+
+
+int num_monotonic_counter_slots(uint16_t counter_desc, uint16_t *counter_slots)
+{
+	const struct monotonic_counter *counter
+			= get_counter_struct(counter_desc);
+	uint16_t num_slots = 0;
+
+	if (counter == NULL || counter_slots == NULL) {
+		return -EINVAL;
+	}
+
+	num_slots = nrfx_nvmc_otp_halfword_read((uint32_t)&counter->num_counter_slots);
+	if (num_slots == 0xFFFF) {
+		/* We consider the 0xFFFF as invalid since it is the default value of the OTP */
+		*counter_slots = 0;
+		return -EINVAL;
+	}
+
+	*counter_slots = num_slots;
+	return 0;
+}
+
+
+/** Function for getting the current value and the first free slot.
+ *
+ * @param[in]   counter_desc  Counter description/home/geva/repos/george_nrf/nrf/subsys/bootloader/bl_otp_counters/bl_otp_counters.c.
+ * @param[out]  counter_vlaue The current value of the requested counter.
+ * @param[out]  free_slot     Pointer to the first free slot. Can be NULL.
+ *
+ * @return The current value of the counter (the highest value before the first
+ *         free slot).
```

Comment:
Why both have an output variable and a return value for the same thing?

Reply:
"(the highest value before the first free slot)."

Wait what, it's not the last used slot that is the current value?

Reply:
Can be dismissed as unrelated as it's the same before the PR isn't it?

Reply:
I will update the return doc since it is describing the old version of the API. I will remove the part that says about the free slots since it is an implementation detail. 

But FYI the highest value before the first free slot and the last used slot is the same slot.